---
title: "SARS-CoV-2 Sequence report KMA AUH"
author: "Marc Trunjer Kusk Nielsen & Carl M. Kobel"
date: "`r Sys.Date()`"
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: console
---

```{r global-options, include = FALSE}
#CHANGE
#knitr::opts_chunk$set(warning = FALSE, message = FALSE, out.width = "100%", dev = c("svg", "png"))
knitr::opts_chunk$set(warning = FALSE, message = FALSE, out.width = "100%")
```


# Introduktion

Denne rapport opsummerer resultaterne af SARS-CoV-2 sekventeringen foretaget af Klinisk Mikrobiologi AUH i samarbejde med MOMA AUH.

Temaer for rapporten:

- Signaturmutationer og deres afledte sandsynlige lineage
- Udvikling i Lineages og Clades
- "Delta"-mutation (Δ69-70 i spike) og dens association til lineages, særligt de bekymrende varianter
- Generel mutationsovervågning over tid
- Svartid
- Ændringer i PCR primer sites
- Kvalitetskontrol

I rapporten anvendes udtrykket 'VOC', det står for Variant of concern, i danske medier omtales de ofte blot som 'Varianter'.

Ugeangivelserne i teksten er ISO uger, som defineret i [ISO 8601](https://en.wikipedia.org/wiki/ISO_week_date). I 2021 medfører det at den første dag i isouge 1 (2021-W01-01) er torsdag 4. januar.


# Metode

Positive SARS-CoV-2 prøver med en CT værdi under 32 sendes til sekventering. Der anvendes amplicon sekventering på Illumina platformen med qiagen kit.

Basecalling foretages af MOMA. Konsensus sekvenser dannes via [ivar](https://github.com/andersen-lab/ivar) pipelinen. Kildekoden til KMA AUH's pipeline (inklusiv denne rapport) er tilgængelig på [github](https://github.com/cmkobel/pipe19).

Generelt forsøger vi at præsentere data med en visualisering og dernæst med en tabel der kan slås op i. 
Tabellerne vil mange steder indeholde automatisk beregnede proportions estimater `point_est` og tilhørende 95% binomial konfidensintervalsgrænser angivet som hhv. `lower`og `upper`.

# Opsætning

Først indlæses hjælpepakker til R og funktioner defineres til brug i analysen.

```{r setup, warning = FALSE, message = FALSE}
library(stats)
library(tidyverse)
library(patchwork)
library(paletteer)
library(ggforce)
library(naniar)
library(here)
library(UpSetR)
theme_set(theme_minimal())

sum_pos <- . %>% 
  summarise(n = n()) %>% 
  mutate(sum = sum(n)) %>% 
  ungroup %>% 
  mutate(binconf = map2(n, sum, ~ Hmisc::binconf(.x, .y) %>% as_tibble)) %>% 
  unnest(binconf) %>% 
  janitor::clean_names() 

table_count <- 0
table_counter <- function() {
  table_count <<- table_count + 1
  paste0("Tabel ", table_count, ". ")
}

make_fancy_caption <- function(fancy_caption = "") {
  htmltools::tags$caption(
                    style = 'caption-side: top; text-align: left;',
                    table_counter(), 
                    htmltools::strong(fancy_caption))
}

sum_pos_table <- function(.data, fancy_caption = "", ...) {
  if(nrow(.data) == 0) {
    return("Ingen sekvenser opfylder selektionskriteriet, se filter i denne kodeblok")
  }
  .data %>% 
    sum_pos %>% 
    mutate(across(c(point_est, lower, upper), scales::percent_format(accuracy = 0.01))) %>% 
    DT::datatable(rownames = FALSE, 
                  filter = "top", 
                  caption = make_fancy_caption(fancy_caption),
                  options = list(pageLength = 5), 
                  ...)
}
  
is_mutant <- function(x) {
  if(is.logical(x)) {
    # if_else(x, "Mutant", "Wildtype", "Inconclusive") %>%
    # factor(levels = c("Inconclusive", "Mutant", "Wildtype"))
    fct_recode(as.character(x), "Mutant" = "TRUE", "Wildtype" = "FALSE") %>% 
      fct_explicit_na("Inconclusive") %>% 
      fct_expand("Wildtype", "Mutant", "Inconclusive")
  } else {
    x
  }
}

`%not_in%` <- Negate(`%in%`)

to_title <- function(name) {
  if(str_detect(name, "[A-Z][0-9]{1,3}[A-Z]")) {
    name
  } else {
    snakecase::to_title_case(name)
  }
}

plot_stacked_area_by <- function(..., .data = ready) {
  groups <- enquos(...)
  
  df <- .data %>% 
  count(...) %>% 
  complete(..., fill = list(n = 0)) %>% 
  group_by(!!groups[[1]]) %>% 
  mutate(prop = n / sum(n)) %>% 
  ungroup 
  
  p1 <- df %>% 
  ggplot(aes(!!groups[[1]], prop, fill = is_mutant(!!groups[[2]]))) +
  geom_area(colour = "black", position = position_stack(reverse = TRUE)) +
  scale_y_continuous("Proportion", labels = scales::percent_format()) +
  labs(x = "", y = "Proportion", fill = to_title(as_label(groups[[2]]))) +
  theme_minimal() +
  theme(legend.position = "bottom")
  
  p2 <- df %>% 
  ggplot(aes(!!groups[[1]], n, fill = is_mutant(!!groups[[2]]))) +
  geom_col(colour = "black", position = position_stack(reverse = TRUE)) +
  labs(x = "", y = "Antal", fill = to_title(as_label(groups[[2]]))) +
  theme_minimal() +
  theme(legend.position = "bottom")
  
if(pull(ready, !!groups[[2]]) %>% is.logical) {
    p1 <- p1 + scale_fill_manual(values = c("Wildtype" = "#76b5c5", "Mutant" = "#873e23", "Inconclusive" = "#cccccc"))
    p2 <- p2 + scale_fill_manual(values = c("Wildtype" = "#76b5c5", "Mutant" = "#873e23", "Inconclusive" = "#cccccc"))
  }
  
  p1 + p2 + plot_layout(guides = 'collect') & theme(legend.position = "bottom")
}
```

Dernæst indlæses og klargøres data til analyse.

```{r illumina, warning = FALSE}
#nanopore <- read_tsv(here("output", "metadata_init.tsv")) %>%

#nanopore = read_tsv("~/GenomeDK/clinmicrocore/pappenheim-receiver/output/metadata_init.tsv") %>% 


# paste("listing files")
# paste(list.files(".."))

nanopore <- read_tsv("metadata_init.tsv") %>% 
#nanopore <- read_tsv("../output/metadata_init.tsv") %>% 
  filter(#batch_id != "batch", # batch_id is now parsed correctly as a numeric column. 
         type == "sample")
```

```{r ready, warning = FALSE}
VOCs <- c("B.1.1.7", "B.1.351", "P.1", "A.23.1", "B.1.525") #ADD NEW MUTATION
MOIs <- # Mutations of interest
          c("S:H69-", "S:V70-", 
          "S:E484K", "S:N501Y", "S:Y453F") #ADD NEW MARKERS

ready_step1 <- nanopore %>% 
  mutate(date  = coalesce(as.Date(modtaget), lubridate::ymd(substr(batch_id, 0, 8))),
         week  = ISOweek::ISOweek(date),
         epi_date_week  = ISOweek::ISOweek2date(paste0(week, "-1")),
         epi_date_month = lubridate::floor_date(date, "month"),
         missing_vector = map(ne_missing, ~ eval(parse(text = paste("c(", gsub("\\-", ":", .x), ")")))) #https://r.789695.n4.nabble.com/convert-delimited-strings-with-ranges-to-numeric-td4673763.html
   ) %>%
  filter(as.integer(ne_totalMissing) < 3000,
         date > lubridate::ymd("2020-12-01"))

ready <- ready_step1 %>% 
  mutate(D614G = str_detect(ne_aaSubstitutions, "S:D614G"),
         N439K = str_detect(ne_aaSubstitutions, "S:N439K"),
         N501Y = str_detect(ne_aaSubstitutions, "S:N501Y"),
         E484K = str_detect(ne_aaSubstitutions, "S:E484K"),
         Y453F = str_detect(ne_aaSubstitutions, "S:Y453F"),
         K417T = str_detect(ne_aaSubstitutions, "S:K417T"),
         P681H = str_detect(ne_aaSubstitutions, "S:P681H"),
         K417T = str_detect(ne_aaSubstitutions, "S:K417T"),
         del69_70 = str_detect(ne_deletions, "21766-21772"), # NA giver problemer - gælder faktisk nok også for de andre, Inkonklusiv bør nok komme fra et qc felt eller et error felt
         missing21766 = any(21766:21772 %in% missing_vector),
         del69_70 = if_else((is.na(del69_70) | del69_70 == FALSE) & missing21766 == FALSE, FALSE, del69_70),
         `B.1.1.7-like` = del69_70 & N501Y & P681H,
         `B.1.351-like` = E484K & N501Y,
         `P.1.1.28-like` = K417T & E484K & N501Y,
         `mink-like` = Y453F & del69_70,
         variants_of_concern = fct_expand(pa_lineage, VOCs) %>% fct_other(keep = VOCs, other_level = "Not a VOC"),
         common_lineages = fct_lump_n(pa_lineage, n = 5),
         focus_lineages  = fct_other(pa_lineage, keep = c(levels(common_lineages), VOCs))
         ) %>% 
  # rowwise() %>% 
  # # mutate(lowest_ct = min(c_across(c(X16, X19, X22, X25, X28, X31, X34, X37, X40, X43, ct)), na.rm = TRUE),
  # #        lowest_ct = na_if(lowest_ct, Inf)) %>% 
  # ungroup() %>% 
  {.}
```

```{r save-ready, warning = FALSE}
#CHANGE?
#saveRDS(ready, file = here("batch", "ready.RDS"))
```

# Lineage og Clade udvikling

Først to figurer der viser udviklingen i Clade og lineage over tid. Bemærk at lineage definitionen er væsentligt mere fingrynet end clade definitionen.

```{r lineage-clade-plots, warning = FALSE}
plot_stacked_area_by(epi_date_week, ne_clade)
plot_stacked_area_by(epi_date_week, pa_lineage)
```

```{r lineage-tables}
ready %>% 
  group_by(pa_lineage) %>% 
  sum_pos_table("Opsummering af lineages")

ready %>% 
  group_by(week, pa_lineage) %>%
  sum_pos_table("Lineages per uge")
```


```{r clade-tables}
ready %>% 
  group_by(ne_clade) %>% 
  sum_pos_table("Opsummering af clades")

ready %>% 
  group_by(week, ne_clade) %>%
  sum_pos_table("Clades per uge")
```

Lineage figuren bliver hurtigt svær at aflæse, derfor vises herunder de 5 hyppigste lineages og de specificerede bekymrende varianter: `r VOCs`

```{r focus-lineages-plot-table, warning=FALSE}
plot_stacked_area_by(epi_date_week, focus_lineages) & scale_fill_paletteer_d("colorblindr::OkabeIto")

ready %>% 
  group_by(week, focus_lineages) %>% 
  sum_pos_table("Lineages der enten er hyppige eller bekymrende, opgjort per uge")
```

# Signaturmutationer og deres afledte lineages

Herunder vises figurer for udvikling i signaturmutationer. 
Signaturmutationerne giver afledte bud på lineages efter følgende skema.... [mangler]

```{r signaturemutation-plots, warning = FALSE}
plot_stacked_area_by(epi_date_week, D614G)
plot_stacked_area_by(epi_date_week, N439K) 
plot_stacked_area_by(epi_date_week, N501Y) 
plot_stacked_area_by(epi_date_week, E484K) 
plot_stacked_area_by(epi_date_week, Y453F)
plot_stacked_area_by(epi_date_week, K417T)
plot_stacked_area_by(epi_date_week, del69_70)
plot_stacked_area_by(epi_date_week, `B.1.1.7-like`)
plot_stacked_area_by(epi_date_week, `B.1.351-like`)
plot_stacked_area_by(epi_date_week, `P.1.1.28-like`)
plot_stacked_area_by(epi_date_week, `mink-like`)
```

Hvor hyppigt observerer vi de enkelte signaturmutationer? 
Første tabel summerer over alle sekvenserne, anden tabel opgør hyppigheden per uge

```{r signaturemutation-tables}
ready %>% 
  summarise(across(D614G:`mink-like`, mean, na.rm = TRUE)) %>% 
  mutate(across(D614G:`mink-like`, scales::percent_format())) %>% 
  DT::datatable(rownames = FALSE, 
                caption = make_fancy_caption("Hyppighed af signaturmutationer"))

ready %>% 
  group_by(week) %>% 
  summarise(across(D614G:`mink-like`, mean, na.rm = TRUE)) %>% 
  mutate(across(D614G:`mink-like`, scales::percent_format(accuracy = 1))) %>% 
  DT::datatable(rownames = FALSE,
                caption = make_fancy_caption("Hyppighed af signaturmutationer, per uge"))
```


# Variants of Concern

Herunder vises to figurer der fokuserer på udviklingen i bekymrings-varianter.
Den første figur viser udviklingen i kontekst af alle sekvenser.
Den anden figur viser udviklingen i kontekst af sekvenser med deletion69-70 i spike. Bemærk at for PANGO lineage B.1.1.7 kan der forekomme sekvenser der *ikke* har deletionen (og således vil være delta-PCR negative).

```{r variants-of-concern-plots, warning = FALSE}
plot_stacked_area_by(epi_date_week, variants_of_concern) & scale_fill_viridis_d(option = "magma")
plot_stacked_area_by(.data = filter(ready, del69_70), epi_date_week, variants_of_concern) & scale_fill_viridis_d(option = "magma") & labs(title = "Kun sekvenser med deletion 69-70")
```

De to næste tabeller gengiver informationen fra de ovenstående figurer.

```{r variants-of-concern-tables, warning = FALSE}
ready %>% 
  group_by(week, variants_of_concern) %>% 
  sum_pos_table("Bekymrende varianter per uge")

ready %>% 
  filter(del69_70) %>% 
  group_by(week, variants_of_concern) %>% 
  sum_pos_table("Bekymrende varianter per uge, i sekvenser med deletion 69-70")
```

I de tre følgende tabeller opgøres hvor godt signaturmutationerne repræsenterer de egentlige lineages.

```{r signaturemutations-versus-voc-tables, warning = FALSE}
count(ready, pa_lineage == "B.1.1.7", `B.1.1.7-like`)
count(ready, pa_lineage == "B.1.351", `B.1.351-like`)
count(ready, pa_lineage == "P.1.1.28", `P.1.1.28-like`)
```


## Epidemikurve for bekmyrende varianter

```{r voc-epicurve}
ready %>% 
  filter(variants_of_concern != "Not a VOC") %>% 
  ggplot(aes(week, colour = variants_of_concern)) +
  stat_count(geom = "point", size = 3) +
  scale_colour_viridis_d(option = "magma") +
  labs(x = "Uge",
       y = "Antal",
       colour = "Variant",
       title = "Epidemikurve for bekymrende varianter")
```

## Ekstra mutationer i VOCs

```{r voc-added-mutations}
ready %>% 
  filter(variants_of_concern != "Not a VOC") %>% 
  separate_rows(ne_aaSubstitutions, sep = ",") %>% 
  count(pa_lineage, ne_aaSubstitutions) %>% 
  arrange(pa_lineage, desc(n)) %>% 
  group_by(pa_lineage) %>% 
  filter(n != max(n)) %>% 
  ungroup() %>% 
  DT::datatable(rownames = FALSE,
                filter = "top",
                caption = make_fancy_caption("Ekstra mutationer i VOC lineages"))
```


# Deletion 69-70 og dens lineages

Tabeller for hver af de tre spørgsmål:
Hvor mange sekvenser har spike deletionen 69-70?
Hvor mange af de sekvenser der har spike deletionen tilhører B.1.1.7?
Hvilke sekvenser tilhører de spike-deletions "positive"? (en tabel der fokuserer på proportioner på tværs af tid og en der fokuserer på rå tal per uge)

```{r deletion-tables, warning = FALSE}
count(ready, del69_70)

ready %>% 
  filter(del69_70) %>% 
  group_by(week) %>% 
  summarise(B.1.1.7 = sum(pa_lineage == "B.1.1.7"), n = n())

ready %>% 
  filter(del69_70) %>% 
  group_by(pa_lineage) %>% 
  sum_pos_table("Lineages, i sekvenser med deletion 69-70")

ready %>% 
  filter(del69_70) %>% 
  select(week, pa_lineage) %>% 
  mutate(present = TRUE) %>% 
  pivot_wider(names_from = pa_lineage, values_from = present, values_fn = length, values_fill = 0) %>% 
  rowwise() %>% 
  mutate(sum_week = sum(c_across(-week))) %>% 
  DT::datatable(rownames = FALSE,
                caption = make_fancy_caption("Antal af prøver med deletion 69-70, per uge og lineage"))
```

# E484K og dens lineages

```{r eeek}
# Vi har ikke nogle E484K, ved 0 rækker giver sum_pos_table en fejl.
ready %>%
  filter(E484K) %>%
  group_by(pa_lineage) %>%
  sum_pos_table("Hvilke lineages har vi med E484K?")

ready %>%
  filter(E484K) %>%
  group_by(week, pa_lineage) %>%
  sum_pos_table("Hvilke lineages har vi med E484K? fordelt på uge")
```

# Mutationsovervågning

## Spike

```{r mutations_long}
mutations_long <- ready %>%
  select(date, full_name, ne_aaSubstitutions) %>%
  separate_rows(ne_aaSubstitutions, sep = ",") %>%
  filter(!is.na(ne_aaSubstitutions)) %>%
  group_by(ne_aaSubstitutions) %>% 
  filter(n() > 1) %>% 
  ungroup %>% 
  mutate(present = TRUE) %>%
  complete(nesting(date, full_name), ne_aaSubstitutions, fill = list(present = FALSE))

mutations_long %>%
  filter(str_detect(ne_aaSubstitutions, "S:")) %>% 
  mutate(present = as.integer(present)) %>%
  ggplot(aes(x = date, y = present)) +
  geom_hline(yintercept = c(0, 0.25, 0.5, 0.75, 1), linetype = "dashed", alpha = 0.5) +
  stat_summary_bin(fun.min = function(x) NA_real_,
                   fun = mean,
                   fun.max = function(x) NA_real_, geom = "line") +
  scale_y_continuous(labels = scales::percent_format(), limits = c(0, 1)) +
  labs(x = "", 
       y = "Proportion", 
       title = "Hyppighed af Spike mutationer", 
       caption = "Mutationer der kun optræder i én sekvens er IKKE inkluderet") +
  facet_wrap(~ ne_aaSubstitutions) +
  theme_void() +
  theme(axis.text.y = element_text())
```

## Spike og øvrige gener

Der medtages kun mutationer der optræder i mindst 5 sekvenser

```{r mutatations-heatmap, dev = "png", fig.width=8}
heatmap_df <- ready %>%
  select(date, full_name, ne_aaSubstitutions) %>%
  separate_rows(ne_aaSubstitutions, sep = ",") %>% 
  group_by(ne_aaSubstitutions) %>% 
  filter(n() > 5) %>% 
  ungroup() %>% 
  mutate(present = 1L) %>% 
  complete(nesting(date, full_name), ne_aaSubstitutions, fill = list(present = 0L)) %>% 
  left_join(select(ready, full_name, epi_date_week, batch_id), by = "full_name") %>% 
  group_by(batch_id, ne_aaSubstitutions) %>% 
  summarise(n_with_substitution = sum(present),
            n_in_batch = n()) %>% 
  mutate(point_est = n_with_substitution / n_in_batch) %>% 
  ungroup %>% 
  janitor::clean_names() %>% 
  mutate(ne_aa_substitutions = fct_reorder(ne_aa_substitutions, point_est),
         batch = as.factor(lubridate::ymd(batch_id)),
         gene = str_extract(ne_aa_substitutions, "[^:]+")) %>%
  {.}
  
  
heatmap_df %>% 
  filter(gene %in% c("N", "S")) %>% 
  ggplot(aes(ne_aa_substitutions, batch_id, fill = point_est)) + 
  geom_raster() + 
  scale_fill_viridis_c(name   = "Proportion",
                       option = "cividis",
                       limits = c(0, 1),
                       trans  = "sqrt") +
  theme_minimal() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) + 
  facet_col(~gene, scales = "free_x") +
  labs(x = "",
       y = "Batch", 
       title = "Hyppighed af mutationer, per gen og batch",
       caption = "Mutationer der optræder i 5 eller færre sekvenser, er IKKE inkluderet")

heatmap_df %>% 
  filter(gene %in% c("ORF1a", "ORF1b")) %>% 
  ggplot(aes(ne_aa_substitutions, batch_id, fill = point_est)) + 
  geom_raster() + 
  scale_fill_viridis_c(name   = "Proportion",
                       option = "cividis", 
                       trans  = "sqrt",
                       limits = c(0, 1)) +
  theme_minimal() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) + 
  facet_col(~gene, scales = "free_x") +
  labs(x = "",
       y = "Batch", 
       title = "Hyppighed af mutationer, per gen og batch",
       caption = "Mutationer der optræder i 5 eller færre sekvenser, er IKKE inkluderet")

heatmap_df %>% 
  filter(gene %in% c("E", "M", "ORF6", "ORF7a", "ORF7b", "ORF8", "ORF9a", "ORF9b")) %>% 
  ggplot(aes(ne_aa_substitutions, batch_id, fill = point_est)) + 
  geom_raster() + 
  scale_fill_viridis_c(name   = "Proportion",
                       option = "cividis", 
                       limits = c(0, 1),
                       trans  = "sqrt") +
  theme_minimal() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  labs(x = "",
       y = "Batch", 
       title = "Hyppighed af mutationer, per gen og batch",
       caption = "Mutationer der optræder i 5 eller færre sekvenser, er IKKE inkluderet")
```

# Mutationskonstellationer

Et foreslået alternativ til at følge lineages og clades over tid er at følge selve mutationerne.
Herunder forsøges at fremstille et aktuelt billede af disse mutations*konstellationer*. Bemærk at det er ikke muligt at medtage alle aktuelle mutationer i figuren.

```{r constellations}
mutations_wide <- ready %>%
  select(full_name, pa_lineage, ne_aaSubstitutions, ne_aaDeletions) %>%
  separate_rows(ne_aaSubstitutions, sep = ",") %>% 
  mutate(sub_present = 1L) %>%
  pivot_wider(names_from = ne_aaSubstitutions, values_from = sub_present, values_fill = 0L) %>% 
  separate_rows(ne_aaDeletions, sep = ",") %>% 
  mutate(del_present = 1L) %>% 
  pivot_wider(names_from = ne_aaDeletions, values_from = del_present, values_fill = 0L) %>% 
  select(-`NA`) %>% # Mange prøver har ingen deletioner, nogle har ikke substitutioner
  as.data.frame() 

mutations_wide %>% 
  upset(nsets = 27, nintersects = 30, order.by = "freq", 
        mb.ratio = c(0.3, 0.7)); grid::grid.text("Alle sekvenser", x = 0.11,  y = 0.965, gp = grid::gpar(fontsize=15))

mutations_wide %>% 
  filter(pa_lineage == "B.1.1.7") %>% 
  upset(nsets = 27, nintersects = 30, order.by = "freq", 
        mb.ratio = c(0.3, 0.7)); grid::grid.text("B.1.1.7", x = 0.05,  y = 0.965, gp = grid::gpar(fontsize=15))

# mutations_wide %>% #Ikke fundet i nanopore prøverne endnu. Evt. aktiver senere
#   filter(pa_lineage %in% c("B.1.177", "B.1.177.12", "B.1.177.21")) %>% 
#   upset(nsets = 27, nintersects = 30, order.by = "freq", 
#         mb.ratio = c(0.3, 0.7)); grid::grid.text("B.1.177*", x = 0.07,  y = 0.965, gp = grid::gpar(fontsize=15))

# mutations_wide %>% 
#   select(any_of(c(VOCs, MOIs))) %>% 
#   upset(nsets = 25, nintersects = 30, order.by = "freq",
#         mb.ratio = c(0.3, 0.7))

# Se på set.metadata argumentet
```

# Ændringer i PCR primer sites

```{r pcr-primer-changes}
ready %>%
  select(date, full_name, ne_pcrPrimerChanges) %>%
  separate_rows(ne_pcrPrimerChanges, sep = ",") %>%
  filter(!is.na(ne_pcrPrimerChanges)) %>%
  mutate(present = TRUE) %>%
  complete(nesting(date, full_name), ne_pcrPrimerChanges, fill = list(present = FALSE)) %>%
  select(full_name, ne_pcrPrimerChanges, present) %>%
  pivot_wider(id_cols = full_name, names_from = ne_pcrPrimerChanges, values_from = present) %>%
  select(-full_name) %>%
  summarise(n = n(), across(everything(), sum)) %>%
  gather(ne_pcrPrimerChange, count, -n) %>%
  arrange(desc(count)) %>%
  mutate(prop = count/n) %>%
  DT::datatable(filter = "top", rownames = FALSE, options = list(pageLength = 25),
                caption = make_fancy_caption("Ændringer i PCR primer sites"))
```

# Svartid

```{r}
svartid <- ready %>% 
  mutate(svartid = lubridate::ymd(batch_id) - modtaget + 1) 

svartid_stats <- svartid %>% 
  summarise(median_svartid = median(svartid, na.rm = TRUE),
            count_at_median = table(svartid)[as.character(median_svartid)])

svartid %>% 
  ggplot(aes(svartid)) + 
  geom_bar() +  
  geom_point(data = svartid_stats, aes(median_svartid, count_at_median), colour = "royalblue4") + 
  geom_text(data = svartid_stats, aes(median_svartid, count_at_median * 1.03, label = "Median"), colour = "royalblue4") +
  scale_x_continuous(breaks = seq(0, 50, by = 10), minor_breaks = seq(0, 50, by = 2)) + 
  theme_minimal() + 
  labs(x = "Dage fra prøvemodtagelse til sekvens",
       y = "Antal",
       title = "Svartid")

svartid %>% 
  ggplot(aes(modtaget, svartid)) + 
  geom_jitter(alpha = 0.2) + 
  geom_smooth(se = FALSE) + 
  scale_y_continuous(breaks = seq(0, 60, by = 10)) +
  coord_cartesian(ylim = c(0, 55)) +
  labs(x = "Prøve modtaget",
       y = "Dage fra prøvemodtagelse til sekvens", 
       caption = "Punkterne er gjort delvist gennemsigtige og skubbet lidt for at visualisere mængden af sekvenser ved hver punkt",
       title = "Udvikling i svartid for sekvensering")
```


# Kvalitetskontrol

Sekvenser med manglende baser >= 3000: `r sum(as.integer(nanopore$ne_totalMissing) >= 3000, na.rm = TRUE)`

```{r qc}
qc <- nanopore %>%
  filter(type == "sample") %>% 
  mutate(ne_qc.overallStatus = factor(ne_qc.overallStatus,
                                   levels = c("bad", "mediocre", "good"),
                                   ordered = TRUE),
         fail_category = cut(ne_totalMissing_interpreted, 
                             breaks = c(0, 300, 600, 3000, 30000),
                             labels = c("0-300",
                                        "300-600",
                                        "600-3000",
                                        "3000-alle")),
         fail_category = fct_rev(fail_category)) %>%
  rowwise() %>%
  mutate(lowest_ct = ct,
         lowest_ct = na_if(lowest_ct, Inf)) %>%
  mutate_at(vars(lowest_ct, ne_totalMissing), as.integer) %>%
  ungroup() %>%
  {.}
```


```{r qc-2}
qc %>% 
  ggplot(aes(lowest_ct, ne_totalMissing_interpreted)) + 
  geom_miss_point(alpha = 0.2, prop_below = 0.55) +
  scale_x_continuous(breaks = seq(0, 40, by = 10), minor_breaks = seq(0, 40, by = 2)) +
  coord_cartesian(xlim = c(-5, 40)) + 
  labs(x = "Ct", y = "Manglende baser", colour = "Manglende data på variablen",
       title = "Manglende baser ifht. Ct",
       caption = "Punkterne er gjort delvist gennemsigtige for at visualisere mængden af sekvenser ved hver punkt") +
  theme(legend.position = "bottom")
```


```{r qc-3}
four_colour_okabe <- c("#E69F00", "#F0E442", "#009E73", "#56B4E9")
qc %>% 
  filter(lowest_ct > 3) %>% 
  filter(!is.na(fail_category)) %>% 
  ggplot(aes(lowest_ct, colour = fail_category)) + 
  stat_ecdf() + 
  scale_colour_manual(values = four_colour_okabe) +
  scale_x_continuous(breaks = seq(10, 40, by = 5), minor_breaks = seq(0, 40, by = 1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.2)) +
  coord_cartesian(xlim = c(10, 40)) + 
  labs(x = "Ct", 
       y = "Proportion i gruppen", 
       title = "Fordeling af sekvenser i kategorier",
       colour = "Sekvenser med x-y manglende baser,\nfærre manglende er bedre") + 
  theme(legend.position = "bottom")
  
qc3b_left <- qc %>% 
  filter(lowest_ct > 3, !is.na(fail_category)) %>% 
  ggplot(aes(lowest_ct, fill = fail_category)) + 
  geom_bar(position = "fill", colour = "black") +
  scale_fill_manual(values = four_colour_okabe) +
  scale_x_continuous(breaks = seq(10, 40, by = 5), minor_breaks = 10:40) +
  coord_cartesian(xlim = c(10, 40)) +
  theme_minimal() +
  labs(x = "Ct",
       y = "Proportion", 
       fill = "Manglende baser",
       title = "Sekvenser med manglende baser ifht. Ct")

qc3b_right <- qc %>% 
  filter(lowest_ct > 3, !is.na(fail_category)) %>% 
  ggplot(aes(lowest_ct, fill = fail_category)) + 
  geom_bar(colour = "black") +
  scale_fill_manual(values = four_colour_okabe) +
  scale_x_continuous(breaks = seq(10, 40, by = 5), minor_breaks = 10:40) +
  coord_cartesian(xlim = c(10, 40)) +
  theme_minimal() +
  labs(x = "Ct",
       y = "Antal", 
       fill = "Manglende baser",
       title = "")

qc3b_left + qc3b_right + plot_layout(guides = "collect") & theme(legend.position = "bottom")
```


```{r qc-4}
qc4_left <- qc %>%
  filter(lowest_ct > 3) %>%
  count(lowest_ct, ne_qc.overallStatus) %>%
  ggplot(aes(lowest_ct, n, fill = ne_qc.overallStatus)) +
  geom_col(position = "fill", colour = "black") +
  scale_fill_paletteer_d("colorblindr::OkabeIto") +
  scale_x_continuous(breaks = seq(10, 40, by = 5), minor_breaks = 10:40) +
  coord_cartesian(xlim = c(10, 40)) +
  theme_minimal() +
  labs(x = "Ct",
       y = "Proportion", 
       fill = "Kvalitet iflg. nextclade",
       title = "Kvalitet ifht. Ct")

qc4_right <- qc %>%
  filter(lowest_ct > 3) %>%
  count(lowest_ct, ne_qc.overallStatus) %>%
  ggplot(aes(lowest_ct, n, fill = ne_qc.overallStatus)) +
  geom_col(colour = "black") +
  scale_fill_paletteer_d("colorblindr::OkabeIto") +
  scale_x_continuous(breaks = seq(10, 40, by = 5), minor_breaks = 10:40) +
  coord_cartesian(xlim = c(10, 40)) +
  theme_minimal() +
  labs(x = "Ct",
       y = "Antal", 
       fill = "Kvalitet iflg. nextclade")

qc4_left + qc4_right + plot_layout(guides = "collect") & theme(legend.position = "bottom")
```


```{r qc-5}
#Nanopore data indeholder ikke information til plottet?
# qc %>% 
#   mutate_at(vars(lowest_ct, ne_totalMissing_interpreted), as.integer) %>% 
#   mutate(eurofins = if_else(`SARS2 Eurofins COV` == "Positiv", "Eurofins", "KMA AUH", missing = "KMA AUH")) %>% 
#   ggplot(aes(ne_totalMissing_interpreted, colour = eurofins)) + 
#   stat_ecdf() + 
#   facet_zoom(xlim = c(0, 3000)) +
#   scale_x_continuous(breaks = c(0, 300, 600, 3000, 5000, 10000, 20000, 30000)) +
#   labs(x = "Manglende baser",
#        y = "Proportion",
#        colour = "Laboratorie") + 
#   theme_gray() +
#   theme(legend.position = "bottom")
```

```{r}
#CODE FOR CT VALUES
# ct <- tibble(nanopore$sample_id, nanopore$ss_ct, as.numeric(nanopore$ss_ct), nanopore$ct)
# ct <- ct %>% 
#   separate_rows(`nanopore$ss_ct`, sep = "/") %>% mutate_at(vars(`nanopore$ss_ct`), as.numeric) %>% 
#   group_by(`nanopore$sample_id`) %>% 
#   summarise(ss_ct = mean(`nanopore$ss_ct`), min_ct = min(ss_ct, `nanopore$ct`))
```

